botorch_community/models/vbll_helper.py

```diff
@@ -0,0 +1,487 @@
+            )
+        elif parameterization == "dense":
+            self.W_logdiag = nn.Parameter(
+                torch.randn(out_features, in_features, dtype=self.dtype)
+                - 0.5 * np.log(in_features)
+            )
+            self.W_offdiag = nn.Parameter(
+                torch.randn(out_features, in_features, in_features, dtype=self.dtype)
+                / in_features
+            )
+        elif parameterization == "dense_precision":
+            self.W_logdiag = nn.Parameter(
+                torch.randn(out_features, in_features, dtype=self.dtype)
+                + 0.5 * np.log(in_features)
+            )
+            self.W_offdiag = nn.Parameter(
+                torch.randn(out_features, in_features, in_features, dtype=self.dtype)
+                * 0.0
+            )
+        elif parameterization == "lowrank":
+            self.W_logdiag = nn.Parameter(
+                torch.randn(out_features, in_features, dtype=self.dtype)
+                - 0.5 * np.log(in_features)
+            )
+            self.W_offdiag = nn.Parameter(
+                torch.randn(out_features, in_features, cov_rank, dtype=self.dtype)
+                / in_features
+            )
+
+    def W(self):
+        cov_diag = torch.exp(self.W_logdiag)
+        if self.W_dist == Normal:
+            cov = self.W_dist(self.W_mean, cov_diag)
+        elif (self.W_dist == DenseNormal) or (self.W_dist == DenseNormalPrec):
+            tril = torch.tril(self.W_offdiag, diagonal=-1) + torch.diag_embed(cov_diag)
+            cov = self.W_dist(self.W_mean, tril)
+        elif self.W_dist == LowRankNormal:
+            cov = self.W_dist(self.W_mean, self.W_offdiag, cov_diag)
+
+        return cov
+
+    def noise(self):
+        return Normal(self.noise_mean, torch.exp(self.noise_logdiag))
+
+    def forward(self, x):
+        out = VBLLReturn(
+            self.predictive(x), self._get_train_loss_fn(x), self._get_val_loss_fn(x)
+        )
+        return out
+
+    def predictive(self, x):
+        return (self.W() @ x[..., None]).squeeze(-1) + self.noise()
+
+    def _get_train_loss_fn(self, x):
+        def loss_fn(y):
+            # construct predictive density N(W @ phi, Sigma)
+            W = self.W()
+            noise = self.noise()
+            pred_density = Normal((W.mean @ x[..., None]).squeeze(-1), noise.scale)
+            pred_likelihood = pred_density.log_prob(y)
+
+            trace_term = 0.5 * (
+                (W.covariance_weighted_inner_prod(x.unsqueeze(-2)[..., None]))
+                * noise.trace_precision
+            )
+
+            kl_term = gaussian_kl(W, self.prior_scale)
+            wishart_term = (
+                self.dof * noise.logdet_precision
+                - 0.5 * self.wishart_scale * noise.trace_precision
+            )
+            total_elbo = torch.mean(pred_likelihood - trace_term)
+            total_elbo += self.regularization_weight * (wishart_term - kl_term)
```

Comment:
in-pace updates can be problematic for backprop, suggest avoiding that
```suggestion
            total_elbo = total_elbo + self.regularization_weight * (wishart_term - kl_term)
```